Issue/5624 new stats header

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes #5624 

⚠️ PR ready to be reviewed but please **do not merge** until @adamzelinski approves the design changes. 

### Description
Adds a new UI for stats **header** on My Store tab. These are the expected designs (iOS based) 

![StatsHeader](https://user-images.githubusercontent.com/2663464/150182490-c25b6a7d-a838-402e-9753-4787637b4e24.png)

A couple things to consider from this new design: 

- Revenue values are not rounded anymore #5653 
- There is no date range bar anymore. Date ranges are now placed inside the scrollable area as part of MyStoreStatsView. This enables removing some classes such as` MyStoreDateRangeView.kt` and unnecessary listener such as `MyStoreStatsListener.kt`
- Removes `dashboard_recency_text` from below the chart as it is not included in the designs anymore.

### Testing instructions
Open My Store Tab and check that chart header values are displayed correctly ad match designs vbzgTKHV9XsyyN5cxxAuhv-fi-1799%3A86756

### Images/gif

https://user-images.githubusercontent.com/2663464/149821776-76a66ebe-3520-4b4a-83af-7aa3024ad8a6.mp4

Some **special cases** to be aware of: 

| Screenshot 1  | Screenshot 2 | Screenshot 3 |
| ------------- | ------------- | ------------- |
| Visitors count is 0, so we can’t calculate conversion rate | For JCP sites, visitor stats are unavailable due to Jetpack, and thus the conversion rate isn’t available  | In the Today tab where hourly visitor stats are unavailable in the API, both visitor and conversion stats are unavailable |
| <img src="https://user-images.githubusercontent.com/2663464/150148870-e9db5936-8c84-453f-ac8e-e90f76842e8e.png" width="200">  | <img src="https://user-images.githubusercontent.com/2663464/150148332-71be8102-1929-4292-959b-9a737315dffb.png" width="200">  | <img src="https://user-images.githubusercontent.com/2663464/150148341-8167c48e-73d3-467d-997e-12174236a9ee.png" width="200"> |



- [X] Release notes have been updated accordinglt in this PR: #5658

JorgeMucientes committed:
Add xml layout for new my store stats header UI

JorgeMucientes committed:
Add logic to set values on new header stats

JorgeMucientes committed:
Set conversion rate and update it accordingly to date granularity

JorgeMucientes committed:
Set pink color for stats header label on value selected from chart

JorgeMucientes committed:
Avoid rounded revenue stats

JorgeMucientes committed:
Remove unused MyStoreDateRangeView.kt

JorgeMucientes committed:
Remove recency message from chart design

JorgeMucientes committed:
Fix detekt issues

JorgeMucientes committed:
Handle cases were visitor number is 0 or unavailable

JorgeMucientes committed:
Remove "Updated moments ago" string from stats empty view

JorgeMucientes committed:
Merge branch 'trunk' into issue/5624-new-stats-header

JorgeMucientes committed:
Handle special cases for visitor stats unavailable

JorgeMucientes commented on WooCommerce/src/main/res/layout/dashboard_main_stats_row.xml:
```diff
@@ -2,122 +2,158 @@
 <LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Better to review as new file

ghost commented:
Looks great @JorgeMucientes awesome work on this. I have a couple of questions:
1. Is there a reason why the '$' is after the amount or is the just the currency that's been selected?
2. I think $0 should be shown as $0 until there's an amount to display ie. $0.50 etc.
3. Are the commas and decimal points in the correct positions or again is this just because of the currency that's being shown?

JorgeMucientes commented:
Thanks for reviewing this @adamzelinski  🙂 . Answering your questions: 
 1. This is due to the configuration I have for my site in `wp-admin` The app will apply the selected WooCommerce settings:
![Screenshot 2022-01-20 at 09 32 01](https://user-images.githubusercontent.com/2663464/150302231-7cc656c6-5fe8-476c-ae69-89ba5d8f001f.png)

2. So If the value is 0 we should remove the decimals right. Makes sense. Going to address that now. 
3. Same as 1. This is dependant on the store site's config.

ghost commented:
Awesome, thanks for clarifying @JorgeMucientes  I'll approve now

JorgeMucientes committed:
Remove empty stats view from Home Screen

JorgeMucientes committed:
Avoid decimals on revenue when value is 0

JorgeMucientes committed:
Fix UI tests after removing dashboard_recency_text from home screen

JorgeMucientes commented on WooCommerce/src/androidTest/kotlin/com/woocommerce/android/screenshots/mystore/StatsComponent.kt:
```diff
@@ -19,12 +17,14 @@ class StatsComponent : Screen {
         // the option doesn't prove to more reliable.
         // idleFor(1000)
-        if (!waitForElementToBeDisplayedWithoutFailure(R.id.dashboard_recency_text)) {
+        if (!waitForElementToBeDisplayedWithoutFailure(R.id.conversionValueTextView)) {
```
`dashboard_recency_text` has been removed in the new designs. This was making the screenshots tests failing. I've replaced it with a new value that should be equivalent in terms of ensuring that that stats are loaded before proceeding with the screenshots.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreStatsView.kt:
```diff
@@ -93,45 +85,41 @@ class MyStoreStatsView @JvmOverloads constructor(
-    private val visitorsValue
-        get() = binding.root.findViewById<MaterialTextView>(R.id.visitors_value)
+    private val statsDateValue
+        get() = binding.root.findViewById<MaterialTextView>(R.id.statsDateTextView)
```
I believe we can simplify this expression to this:
```suggestion
        get() = binding.statsViewRow.statsDateTextView
```

Same applies to the properties below.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreStatsView.kt:
```diff
@@ -297,38 +355,56 @@ class MyStoreStatsView @JvmOverloads constructor(
-        revenueValue.setTextColor(color)
-        ordersValue.setTextColor(color)
+    private fun updateConversionRate() {
+        val ordersCount = ordersValue.text.toString().toIntOrNull()
+        val visitorsCount = visitorsValue.text.toString().toIntOrNull()
+
+        if (visitorsCount == null || ordersCount == null) {
+            conversionValue.isVisible = false
+            binding.statsViewRow.emptyConversionRateIndicator.isVisible = true
+            return
+        }
+
+        val conversionRateDisplayValue = when (visitorsCount) {
+            0 -> "0%"
+            else -> {
+                val conversionRate = (ordersCount / visitorsCount.toFloat()) * 100
+                DecimalFormat("##.#").format(conversionRate) + "%"
+            }
+        }
+        val color = ContextCompat.getColor(context, R.color.color_on_surface_high)
+        conversionValue.setTextColor(color)
+        binding.statsViewRow.emptyConversionRateIndicator.isVisible = false
+        conversionValue.isVisible = true
+        conversionValue.text = conversionRateDisplayValue
```
Shouldn't this logic be in the ViewModel to be testable? I mean the uiModel in the ViewModel should contain a field for conversion rate that's calculated there, and not based on the EditText changes 🤔

hichamboushaba replied:
I know a change like this can be big, so if you agree with my thinking, you can log an issue to improve it later 🙂.

hichamboushaba commented on WooCommerce/src/main/res/layout/wc_empty_stats_view.xml:
```diff
@@ -11,29 +11,11 @@
+            android:layout_marginBottom="@dimen/major_100" />
 
     </LinearLayout>
 </merge>
```
I didn't find any usage for this file, and the app builds fine without it, maybe we can delete it?

JorgeMucientes replied:
Nice catch! Removed

hichamboushaba reviewed (approved):
Nice work @JorgeMucientes, I left some minor comments, but nothing blocking, I'm pre-approving.

JorgeMucientes committed:
Update tabs text color to medium

JorgeMucientes committed:
Simplify expressions to access text views from MyStoreStatsView

JorgeMucientes committed:
Remove unused layout file

JorgeMucientes commented:
Thanks for reviewing this @hichamboushaba I've addressed most of the suggestions :) 

Regarding moving the logic from `updateConversionRate` function to a ViewModel, while I agree that it would be nice, specially because that would make it easily testable, I have doubts as I believe it would make the code more complex for the following reasons: 

- Only this part of the code from function`updateConversionRate` could  be moved to the viewModel: 
``` kotlin
        val conversionRateDisplayValue = when (visitorsCount) {
            0 -> "0%"
            else -> {
                val conversionRate = (ordersCount / visitorsCount.toFloat()) * 100
                DecimalFormat("##.#").format(conversionRate) + "%"
            }
        }
```
The rest is needed inside `updateConversionRate` because we need to update styles, visibilities and values for conversion rate when swiping/dragging over the line chart.  

- Setting listeners to react to Textviews changes, and then update conversion rate, is kind of hacky I know, but it also makes it so easy to cover all cases in a single place. Dealing with this inside the viewModel will require adding several checks in different places to emit a new UIState once VisitorsStatsUI and RevenueStatsUI have been updated.  I mean is doable but I'm not so sure it is worth the refactor just to get rid of this simple block: 

```kotlin 
        val conversionRateDisplayValue = when (visitorsCount) {
            0 -> "0%"
            else -> {
                val conversionRate = (ordersCount / visitorsCount.toFloat()) * 100
                DecimalFormat("##.#").format(conversionRate) + "%"
            }
        }
```

Let me know your troughs, and maybe we can open the ticket and make a deeper analysis on pros and cons. But in anycase I won't apply this refactor in this PR

hichamboushaba commented:
@JorgeMucientes I'm not very familiar with the code here, so maybe my remark wasn't very accurate, so let's merge this :shipit:, and whatever you decide next is good for me 👍.
